File: Python/3 - Linked Lists - Singly & Doubly Linked - Greg Hogg DSA Course Materials Lecture 3.py
# Linked Lists - Singly & Doubly Linked - Greg Hogg DSA Course Materials Lecture 3

# The Colab Notebook can be viewed here: https://colab.research.google.com/drive/1spa5_AB-IXpdZjMeGZ6ARX_LEKRt6bBx


# Singly Linked Lists

class SinglyNode:

  def __init__(self, val, next=None):
    self.val = val
    self.next = next

  def __str__(self):
    return str(self.val)

Head = SinglyNode(1)
A = SinglyNode(3)
B = SinglyNode(4)
C = SinglyNode(7)

Head.next = A
A.next = B
B.next = C

print(Head)

# Traverse the list - O(n)
curr = Head

while curr:
  print(curr)
  curr = curr.next

# Diplay linked list - O(n)
def display(head):
  curr = head
  elements = []
  while curr:
    elements.append(str(curr.val))
    curr = curr.next
  print(' -> '.join(elements))

display(Head)

# Search for node value - O(n)
def search(head, val):
  curr = head
  while curr:
    if val == curr.val:
      return True
    curr = curr.next

  return False

search(Head, 7)

# Doubly Linked Lists

class DoublyNode:
  def __init__(self, val, next=None, prev=None):
    self.val = val
    self.next = next
    self.prev = prev

  def __str__(self):
    return str(self.val)

head = tail = DoublyNode(1)
print(tail)

# Display - O(n)
def display(head):
  curr = head
  elements = []
  while curr:
    elements.append(str(curr.val))
    curr = curr.next
  print(' <-> '.join(elements))

display(head)

# Insert at beginning - O(1)
def insert_at_beginning(head, tail, val):
  new_node = DoublyNode(val, next=head)
  head.prev = new_node
  return new_node, tail

head, tail = insert_at_beginning(head, tail, 3)
display(head)

# Insert at end - O(1)
def insert_at_end(head, tail, val):
    new_node = DoublyNode(val, prev=tail)
    tail.next = new_node
    return head, new_node

head, tail = insert_at_end(head, tail, 7)
display(head)

File: JS/3 - Linked Lists - Singly & Doubly Linked - Greg Hogg DSA Course Materials Lecture 3.js
// Linked Lists - Singly & Doubly Linked - Greg Hogg DSA Course Materials Lecture 3


// Singly Linked Lists
class SinglyNode {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }

    toString() {
        return this.val.toString();
    }
}

class SinglyLinkedList {
    constructor() {
        this.head = null;
    }

    // Add a new node at the end
    append(val) {
        if (!this.head) {
            this.head = new SinglyNode(val);
        } else {
            let current = this.head;
            while (current.next) {
                current = current.next;
            }
            current.next = new SinglyNode(val);
        }
    }

    // Display linked list - O(n)
    display() {
        let elements = "";
        let current = this.head;
        while (current) {
            elements += current.val + " -> ";
            current = current.next;
        }
        elements += "null";
        console.log("Singly linked list: " + elements);
    }

    // Search for node value - O(n)
    search(val) {
        let current = this.head;
        while (current) {
            if (current.val === val) {
                return true;
            }
            current = current.next;
        }
        return false;
    }
}

// Doubly Linked Lists
class DoublyNode {
    constructor(val) {
        this.val = val;
        this.next = null;
        this.prev = null;
    }

    toString() {
        return this.val.toString();
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    // Insert at beginning - O(1)
    insertAtBeginning(val) {
        const newNode = new DoublyNode(val);
        if (this.head) {
            newNode.next = this.head;
            this.head.prev = newNode;
        }
        this.head = newNode;
        if (!this.tail) {
            this.tail = newNode;
        }
    }

    // Insert at end - O(1)
    insertAtEnd(val) {
        const newNode = new DoublyNode(val);
        if (this.tail) {
            this.tail.next = newNode;
            newNode.prev = this.tail;
        }
        this.tail = newNode;
        if (!this.head) {
            this.head = newNode;
        }
    }

    // Display doubly linked list - O(n)
    display() {
        let elements = "";
        let current = this.head;
        while (current) {
            elements += current.val + " <-> ";
            current = current.next;
        }
        elements += "null";
        console.log("Doubly linked list: " + elements);
    }
}

// Example Usage
const singlyList = new SinglyLinkedList();
singlyList.append(1);
singlyList.append(3);
singlyList.append(4);
singlyList.append(7);
console.log("Head of the list: " + singlyList.head);
singlyList.display();

const searchValue = 7;
const found = singlyList.search(searchValue);
console.log("Search for value " + searchValue + " in list: " + found);

const doublyList = new DoublyLinkedList();
doublyList.insertAtBeginning(1);
console.log("Head of the doubly linked list: " + doublyList.head);
doublyList.display();

doublyList.insertAtBeginning(3);
doublyList.display();

doublyList.insertAtEnd(7);
doublyList.display();
